File: DevelopmentUtility/youyou_Proto/CreateLuaProto/TestProto.lua
--测试协议
TestProto = { ProtoCode = 5001, IsSuccess = false, ErrorCode = 0, Name = "", Count = 0, ItemIdTable = { }, RoleTable = { } }

--这句是重定义元表的索引，就是说有了这句，这个才是一个类
TestProto.__index = TestProto;

function TestProto.New()
    local self = { }; --初始化self
    setmetatable(self, TestProto); --将self的元表设定为Class
    return self;
end


--定义角色
Role = { RoleId = 0, RoleName = "" }
Role.__index = Role;
function Role.New()
    local self = { };
    setmetatable(self, Role);
    return self;
end


--发送协议
function TestProto.SendProto(proto)

    local ms = CS.LuaHelper.Instance:CreateMemoryStream();
    ms:WriteUShort(proto.ProtoCode);

    ms:WriteBool(proto.IsSuccess);
    if(proto.IsSuccess) then
        ms:WriteUTF8String(Name);
        else
        ms:WriteInt(ErrorCode);
    end
    ms:WriteInt(proto.Count);
    for i = 1, proto.Count, 1 do
        ms:WriteInt(ItemIdList[i]);
        ms:WriteInt(RoleList[i].RoleId);
        ms:WriteUTF8String(RoleList[i].RoleName);
    end

    CS.LuaHelper.Instance:SendProto(ms:ToArray());
    ms:Dispose();
end


--解析协议
function TestProto.GetProto(buffer)

    local proto = TestProto.New(); --实例化一个协议对象
    local ms = CS.LuaHelper.Instance:CreateMemoryStream(buffer);

    proto.IsSuccess = ms:ReadBool();
    if(proto.IsSuccess) then
        proto.Name = ms:ReadUTF8String();
        else
        proto.ErrorCode = ms:ReadInt();
    end
    proto.Count = ms:ReadInt();
	proto.ItemIdTable = {};
	proto.RoleTable = {};
    for i = 1, proto.Count, 1 do
        local _ItemId = ms:ReadInt();  --物品ID
        proto.ItemIdTable[#proto.ItemIdTable+1] = _ItemId;
        local _Role = Role.New();
        _Role.RoleId = ms:ReadInt();
        _Role.RoleName = ms:ReadUTF8String();
        proto.RoleTable[#proto.RoleTable+1] = _Role;
    end

    ms:Dispose();
    return proto;
end

File: DevelopmentUtility/youyou_Proto/CreateProto/TestProto.cs
//===================================================
//作    者：
//创建时间：2022-03-22 14:16:18
//备    注：
//===================================================
using System.Collections;
using System.Collections.Generic;
using System;

/// <summary>
/// 测试协议
/// </summary>
public struct TestProto : IProtocol
{
    public ushort ProtoCode { get { return 5001; } }

    public bool IsSuccess; //是否成功
    public int ErrorCode; //错误编号
    public string Name; //名称
    public int Count; //数量
    public List<int> ItemIdList; //物品ID
    public List<Role> RoleList; //角色

    /// <summary>
    /// 角色
    /// </summary>
    public struct Role
    {
        public int RoleId; //角色ID
        public string RoleName; //角色名称
    }

    public byte[] ToArray()
    {
        using (MMO_MemoryStream ms = new MMO_MemoryStream())
        {
            ms.WriteUShort(ProtoCode);
            ms.WriteBool(IsSuccess);
            if(IsSuccess)
            {
                ms.WriteUTF8String(Name);
            }
            else
            {
                ms.WriteInt(ErrorCode);
            }
            ms.WriteInt(Count);
            for (int i = 0; i < Count; i++)
            {
                ms.WriteInt(ItemIdList[i]);
                ms.WriteInt(RoleList[i].RoleId);
                ms.WriteUTF8String(RoleList[i].RoleName);
            }
            return ms.ToArray();
        }
    }

    public static TestProto GetProto(byte[] buffer)
    {
        TestProto proto = new TestProto();
        using (MMO_MemoryStream ms = new MMO_MemoryStream(buffer))
        {
            proto.IsSuccess = ms.ReadBool();
            if(proto.IsSuccess)
            {
                proto.Name = ms.ReadUTF8String();
            }
            else
            {
                proto.ErrorCode = ms.ReadInt();
            }
            proto.Count = ms.ReadInt();
            proto.ItemIdList = new List<int>();
            proto.RoleList = new List<Role>();
            for (int i = 0; i < proto.Count; i++)
            {
                int _ItemId = ms.ReadInt();  //物品ID
                proto.ItemIdList.Add(_ItemId);
                Role _Role = new Role();
                _Role.RoleId = ms.ReadInt();
                _Role.RoleName = ms.ReadUTF8String();
                proto.RoleList.Add(_Role);
            }
        }
        return proto;
    }
}